shaping/wrapping.go

```diff
@@ -412,17 +500,27 @@ func (l *LineWrapper) Prepare(config WrapConfig, paragraph []rune, shapedRuns ..
 // It is equivalent to iteratively invoking WrapLine with a constant maxWidth.
 // If the config has a non-zero TruncateAfterLines, WrapParagraph will return at most
 // that many lines. The truncated return value is the count of runes truncated from
-// the end of the text.
-func (l *LineWrapper) WrapParagraph(config WrapConfig, maxWidth int, paragraph []rune, shapedRuns ...Output) (_ []Line, truncated int) {
-	if len(shapedRuns) == 1 && shapedRuns[0].Advance.Ceil() < maxWidth && !(config.TextContinues && config.TruncateAfterLines == 1) {
-		return []Line{shapedRuns}, 0
+// the end of the text. If non-nil, the contents of the provided lines slice will be
+// overwritten with the new lines until it runs out of space, avoiding allocations.
+func (l *LineWrapper) WrapParagraph(config WrapConfig, maxWidth int, paragraph []rune, runs RunIterator, scratch WrapScratch) (_ []Line, truncated int) {
+	lines := scratch.Lines[:0]
+	if _, nextRun, ok := runs.Peek(); ok && runs.IsFinal() && nextRun.Advance.Ceil() < maxWidth && !(config.TextContinues && config.TruncateAfterLines == 1) {
+		if cap(scratch.Line.Line) > 0 {
+			lines = append(lines, scratch.Line.Line[:1])
+			lines[0][0] = nextRun
+		} else {
+			lines = append(lines, []Output{nextRun})
+		}
+		return lines, 0
```

Comment:
Shouldn't we consume the Output, and not only Peek it ?
If so, by calling Peek again, the IsFinal could probably be avoided

Reply:
In hindsight, I think you're right that `IsFinal` can be eliminated here. We can Save() the iterator before doing this check, advance until we know whether there's only one element, and then Restore() if we didn't take that code path.

Reply:
I've took the liberty of adding one (failing) test `TestWrapping_oneLine` to clarify what I mean't. Feel free to remove it if you think it does not display the expected behavior.

Reply:
Thanks for that. It's addressed and passing now.